### PR TITLE
Implement repeatUntil and repeatUntilM

### PR DESCRIPTION
**Pull request by YuvalItzchakov**
This PR adds two combinators to the API surface. These combinators came up as useful during John's ZIO Workshop and I thought it was worth implementing

- `repeatUntil`: Takes a `PartialFunction[A, B]` and repeats until the effect is in the codomain of the partial function
- `repeatUntilM`: Same as `repeatUntil` but takes an effect instead, evaluating it until it yields a success.

### CODE REVIEW

**Comment by ghostdogpr** [comment]
Can you rebase to the latest master? The error you have with the CI has been fixed.

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -1054,6 +1054,24 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     schedule.initial.flatMap(loop(None, _))
   }
 
+  /**
+   * Repeats this effect until the result exists in the partial function
+   * codomain.
```
Comment (neko-kai) [inline core/shared/src/main/scala/zio/ZIO.scala]: I'd probably reword this as "until the partial function succeeds on the result"

Comment (YuvalItzchakov) [reply core/shared/src/main/scala/zio/ZIO.scala]: Yup, done.

**Comment by YuvalItzchakov** [comment]
Should these methods be implemented on the ZIO companion as well?

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 2]
```diff
@@ -1054,6 +1054,23 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
+      f.lift(a) match {
+        case Some(value) => ZIO.succeed(value)
+        case None        => repeatUntil(f)
+      }
+    }
+
+  /**
+   * Effectful version of `repeatUntil`. Will repeat until `f` succeeds
+   */
+  final def repeatUntilM[R1 <: R, E2, B](f: A => ZIO[R1, E2, B]): ZIO[R1, E2, B] =
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZIO.scala]: Should the result be typed `ZIO[R1, Nothing, B]`? The function will never fail.

Also, this is essentially `flatMap(f).eventually`. Does this justify a dedicated combinator?

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 3]
```diff
@@ -1054,6 +1054,23 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     schedule.initial.flatMap(loop(None, _))
   }
 
+  /**
+   * Repeats this effect until the partial function succeeds on the result.
+   */
+  final def repeatUntil[B](f: PartialFunction[A, B]): ZIO[R, E, B] =
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZIO.scala]: Nice

File: core/shared/src/main/scala/zio/ZIO.scala (hunk 2 above)
Comment (YuvalItzchakov) [reply core/shared/src/main/scala/zio/ZIO.scala]: It should. The main use case was using the console to read input and make sure it's valid:

```scala
for {
    _ <- putStrLn("Please enter a time in 24 hour format (hh:mm): ")
    userTime <- getStrLn.repeatUntil(in => Task.effect(in.toString).tapError(_ => putStrLn("Please enter a valid value"))
  } yield userTime
```

Does this justify a dedicated combinator? I'm not sure :)

Comment (iravid) [reply core/shared/src/main/scala/zio/ZIO.scala]: Ok, let's wait for another opinion :-)

Comment (YuvalItzchakov) [reply core/shared/src/main/scala/zio/ZIO.scala]: I talked with @jdegoes about this during the workshop, perhaps he can add his 2 cents.

Comment (jdegoes) [reply core/shared/src/main/scala/zio/ZIO.scala]: The return type should definitely omit the final error, and we can implement it using `(self >>= f).eventually`. 

The symmetry with `repeatUntil` is nice, although perhaps we could make it closer still by allowing `PartialFunction[A, ZIO[R1, E2, B]]`. The implementation of this one is also more complex and possibly justifying its existence.

Thoughts?

Comment (YuvalItzchakov) [reply core/shared/src/main/scala/zio/ZIO.scala]: SGTM. I've reworked the implementation, WDYT now? @iravid @jdegoes @neko-kai

Comment (iravid) [reply core/shared/src/main/scala/zio/ZIO.scala]: Did you mean to convert `repeatUntilM` to take a `PartialFunction`?

Comment (iravid) [reply core/shared/src/main/scala/zio/ZIO.scala]: I.e. `repeatUntil` just uses a plain partial function, `repeatUntilM` allows evaluating an effectful partial function

Comment (YuvalItzchakov) [reply core/shared/src/main/scala/zio/ZIO.scala]: @iravid If I understood correctly, I think he meant for `repeatUntil` to take a `PartialFunction[A, ZIO[R1, E1, B]]` and for `repeatUntilM` to take a total function `A => ZIO[R1, E1, B]`

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 4]
```diff
@@ -1054,6 +1054,23 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
+      f.lift(a) match {
+        case Some(value) => value
+        case None        => repeatUntil(f)
+      }
+    }.eventually
+
+  /**
+   * Effectful version of `repeatUntil`. Will repeat until `f` succeeds
+   */
+  final def repeatUntilM[R1 <: R, E1 >: E, B](f: A => ZIO[R1, E1, B]): ZIO[R1, Nothing, B] =
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/ZIO.scala]: Sorry for the confusion, I think this one should be `PartialFunction[A, ZIO[R1, E1, B]]`, and that `repeatUntil` should be `PartialFunction[A, B]`. Then there is symmetry between them and it matches the overall ZIO convention on the `M` suffix. Does that make sense?

**Comment by jdegoes** [comment]
@YuvalItzchakov This one just needs a small fix and we can merge it. However, I will close the PR now to avoid clutter, please re-open when it makes sense for you to resume!
